src/wallet/wallet.cpp

```diff
@@ -1096,70 +1133,296 @@ bool CWallet::LoadToWallet(const uint256& hash, const UpdateWalletTxFn& fill_wtx
+
+        senderPubKey = XOnlyPubKey(pubkey);
+
+    }
+
+    else if (whichType == TxoutType::WITNESS_V0_KEYHASH || scriptPubKey.IsPayToScriptHash()) {
+        if (scriptWitness.stack.size() != 2) return false;
+        assert(scriptWitness.stack.at(1).size() == 33);
+
+        CPubKey pubkey = CPubKey(scriptWitness.stack.at(1));
+        assert(pubkey.IsFullyValid());
+
+        senderPubKey = XOnlyPubKey(pubkey);
+    }
+
+    else if (whichType == TxoutType::WITNESS_V1_TAPROOT) {
+
+        senderPubKey = XOnlyPubKey(solutions[0]);
+        assert(senderPubKey.IsFullyValid());
+    }
+
+    CTxDestination address;
+    ExtractDestination(scriptPubKey, address);
+
+    return true;
+}
+
+bool CWallet::VerifySilentPayment(const CTransaction& tx, std::vector<std::tuple<CKey, int32_t>>& rawTrKeys)
+{
+    AssertLockHeld(cs_wallet);
+
+    if (tx.IsCoinBase() || tx.vin.empty()) return false;
+
+    if (!IsWalletFlagSet(WALLET_FLAG_DESCRIPTORS)) return false;
+
+    std::vector<std::tuple<CScript, XOnlyPubKey>> outputPubKeys;
+
+    for (auto& vout : tx.vout) {
+
+        if (IsMine(vout)) {
+            continue;
         }
+
+        std::vector<std::vector<unsigned char>> solutions;
+        TxoutType whichType = Solver(vout.scriptPubKey, solutions);
+
+        // Silent Payments require that the recipients use Taproot address
+        if (whichType != TxoutType::WITNESS_V1_TAPROOT) {
+            continue;
+        }
+
+        auto xOnlyPubKey = XOnlyPubKey(solutions[0]);
+
+        assert(xOnlyPubKey.IsFullyValid());
+
+        outputPubKeys.emplace_back(vout.scriptPubKey, xOnlyPubKey);
     }
-    return false;
+
+    if (outputPubKeys.empty()) {
+        return false;
+    }
+
+    std::vector<XOnlyPubKey> sender_pubkeys;
+
+    for (const CTxIn& txin : tx.vin) {
+        auto senderPubKey = XOnlyPubKey();
+
+        if (!ExtractPubkeyFromInput(txin, senderPubKey)) {
+            continue;
+        }
+
+        assert(senderPubKey.IsFullyValid());
+
+        sender_pubkeys.push_back(senderPubKey);
+    }
+
+    XOnlyPubKey sum_sender_pubkeys;
+
+    if (sender_pubkeys.empty() || sender_pubkeys.size() != tx.vin.size()) {
+        if (!m_chain->getSilentTransactionFirstInputPubKey(tx.GetHash(), sum_sender_pubkeys)) {
+            return false;
+        }
+    } else {
+        sum_sender_pubkeys = silentpayment::Recipient::SumXOnlyPublicKeys(sender_pubkeys);
+    }
+
+    for (ScriptPubKeyMan* spkm : GetActiveScriptPubKeyMans()) {
+        DescriptorScriptPubKeyMan* desc_spkm = dynamic_cast<DescriptorScriptPubKeyMan*>(spkm);
+
+        std::string desc_str;
+        bool res_get_desc = desc_spkm->GetDescriptorString(desc_str, false);
+
+        // There must be only one SP descriptor
+        if (res_get_desc && desc_str.rfind("sp(", 0) != 0) {
+            continue;
+        }
+
+        rawTrKeys = desc_spkm->VerifySilentPaymentAddress(outputPubKeys, sum_sender_pubkeys);
+    }
+
+    return rawTrKeys.size() > 0;
```

Comment:
47cc799b9518c13d96fde5a7de050b8a0cd41e4c
```suggestion
    return !rawTrKeys.empty();
```

Reply:
Done in https://github.com/bitcoin/bitcoin/pull/24897/commits/43503e2ae6a462472f8ede023d6682cff99dd60d.